Split partitioned-heat-conduction case into complex and simple one

BenjaminRodenberg opened the pull request:
closes #147

# Specification for `partitioned-heat-conduction`:

* Dirichlet is always left, Neumann is always right
* Coupling Boundary is always straight line parallel to y axis
* No time dependent heat flux and rhs
* `mpirun` is still supported (for `fenics`)
* no subcycling. Timestep of both FEniCS solvers is hard-coded to be equal to time window size from `precice-config.xml`.
* Analytical solution is always possible
* use scalar-valued flux (important for compatibility with nutils and again easier)

# Specification for `partitioned-heat-conduction-complex`:

* Dirichlet and Neumann solvers are arbitrary
* Coupling Boundary is arbitrary
* Time dependent heat flux and rhs may be set via `gamma`
* `mpirun` is supported (for `fenics`)
* subcycling possible
* Analytical solution is possible for straight and circular boundary as long as no subcycling is used.
* use vector-valued flux

# Open Todos

- [x] @IshaanDesai @uekerman Is anything missing or too much in the specification above?
- [x] Need to adapt the structure to #125 
- [x] Make sure that solver can be reused with other compatible cases, as well. (Nutils)
- [x] Make sure that nutils also recovers exact solution (postponed to #152 & #153)

BenjaminRodenberg commented:
@uekerman I assume that `flow-over-heated-plate` is independent from this PR? The FEniCS solver used in `flow-over-heated-plate` is very similar, but has some differences and cannot easily be reused.

uekerman commented:
> @uekerman I assume that `flow-over-heated-plate` is independent from this PR? The FEniCS solver used in `flow-over-heated-plate` is very similar, but has some differences and cannot easily be reused.

yes, `flow-over-heated-plate` is completely independent of this one.

uekerman commented:
Specification looks good so far. I guess a real review only makes sense once we have the compatibility with Nutils.
A comment already now: for that we will need a scalar flux and a nearest-neighbor mapping.

BenjaminRodenberg commented:
> Specification looks good so far. I guess a real review only makes sense once we have the compatibility with Nutils.
> A comment already now: for that we will need a scalar flux and a nearest-neighbor mapping.

To make sure I understand: We want this case for FEniCS and for nutils. Meaning, according to the new structure we get:

* `partitioned-heat-conduction/`
    * `solid-fenics/...`
        * `heat.py`
    * `solid-nutils/...`
        * `heat.py`
    * `precice-config.xml`
    * `...`

This means, that we need a new tutorial (according to the old structure), `precice/tutorials/tree/master/HT/partitioned-heat/nutils-nutils`, first. Does this already exist somewhere? The closest one I can find is [`flow-over-plate/buoyantPimpleFoam-nutils`](https://github.com/precice/tutorials/tree/master/CHT/flow-over-plate/buoyantPimpleFoam-nutils/Nutils).

I think the cleanest approach would be to first create another PR from `develop`, where I collect all the necessary parts from `partitioned-heat/fenics-fenics` and `flow-over-plate/buoyantPimpleFoam-nutils` to develop a new case `partitioned-heat/nutils-nutils`. Afterwards I can use the resulting case, add it in this PR and move everything according to the new structure.

uekerman commented:
> This means, that we need a new tutorial (according to the old structure), precice/tutorials/tree/master/HT/partitioned-heat/nutils-nutils, first. Does this already exist somewhere?

Yes. Sources are provided in the issue: https://github.com/precice/tutorials/issues/137

> To make sure I understand: We want this case for FEniCS and for nutils. Meaning, according to the new structure we get:

The in  https://github.com/precice/tutorials/issues/137 suggested structure is

```
- dirichlet-fenics
  - heat.py
- neumann-fenics
  - heat.py
- dirichlet-nutils
  - heat.py
- neumann-nutils
  - heat.py 
- precice-config.xml
- clean.sh
- README.md
```

Yes, the drawback of this scheme is that it leads to quite some code duplication. Maybe it is simple to understand though.

Alternatively:

```
- fenics
  - heat.py
- nutils
  - heat.py
- precice-config.xml 
```
together with some command line argument to decide whether the left Dirichlet or the right Neumann sub domain should be computed.

BenjaminRodenberg commented:
> > This means, that we need a new tutorial (according to the old structure), precice/tutorials/tree/master/HT/partitioned-heat/nutils-nutils, first. Does this already exist somewhere?
> 
> Yes. Sources are provided in the issue: #137

I added the sources to this PR. Currently just copy-paste without any testing or updates of the code.

> > To make sure I understand: We want this case for FEniCS and for nutils. Meaning, according to the new structure we get:
> 
> The in #137 suggested structure is
> 
> ```
> - dirichlet-fenics
>   - heat.py
> - neumann-fenics
>   - heat.py
> - dirichlet-nutils
>   - heat.py
> - neumann-nutils
>   - heat.py 
> - precice-config.xml
> - clean.sh
> - README.md
> ```
> 
> Yes, the drawback of this scheme is that it leads to quite some code duplication. Maybe it is simple to understand though.

That's probably a good reason not to use this structure here.

> Alternatively:
> 
> ```
> - fenics
>   - heat.py
> - nutils
>   - heat.py
> - precice-config.xml 
> ```
> 
> together with some command line argument to decide whether the left Dirichlet or the right Neumann sub domain should be computed.

I like this idea. Added to this PR :+1:

BenjaminRodenberg commented:
I took care of updating the nutils case as far as I can get. I currently see one problem:

Nutils uses two meshes and FEniCS only one. This does not allow us to use the same configuration for both cases. @uekerman any idea what we can do here?

uekerman commented:
> Nutils uses two meshes and FEniCS only one. This does not allow us to use the same configuration for both cases. @uekerman any idea what we can do here?

We could read and write at the GPs only, that should be simple

BenjaminRodenberg commented:
The last problem with this PR that I see is that the exact solution cannot be recovered when using nutils. Remaining features are fine.

BenjaminRodenberg commented:
I decided to postpone fixing the nutils case to #152 and #153. This PR is from my point of view ready to merge. We can continue working on #153 at a later point in time from my point of view and I don't want with merging the features that are already working until the nutils case is fixed, since I'm a bit stuck here currently and it is hard to estimate how long fixing this will take.

BenjaminRodenberg commented:
Notes for review:

* Are the specifications from above fulfilled?
* Are there still any features or files we don't need for the respective cases?
* Is the documentation clear (enough)?

IshaanDesai commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
+
+* `fenics`, requires you to install [FEniCS](https://fenicsproject.org/download/) and the [FEniCS-adapter](https://github.com/precice/fenics-adapter). 
+
+* :construction: This case is still under construction. See https://github.com/precice/tutorials/issues/152. :construction: `nutils`, requires you to install [Nutils] 
```
Would a link to Nutils documentation or any other code reference be helpful here?

BenjaminRodenberg replied:
Done. https://github.com/precice/tutorials/commit/35dc013342209c62cc53831410e620f4a7cf4063

IshaanDesai reviewed (approved):
Looks good :+1: The separation of *simple* and *complex* geometry cases makes the FEniCS code much more readable :)

BenjaminRodenberg commented:
@uekerman another hint for the review: Currently we don't provide `fenics/run.sh`, since we need command line input to decide whether we run the Dirichlet or Neumann solver. I see two options here:

1. provide a `fenics/run.sh` that accepts command line input
2. provide a `dirichlet-fenics/run.sh` and `neumann-fenics/run.sh`. Put all actual code only into `dirichlet-fenics`. `neumann-fenics` only exists for following the general structure and providing a `run.sh` script.

Any opinion?

uekerman commented on partitioned-heat-conduction-complex/precice-config.xml:
```diff
@@ -0,0 +1,61 @@
+    <data:scalar name="Temperature"/>
+    <data:vector name="Flux"/>
+
+    <mesh name="DirichletNodes">
```
```suggestion
    <mesh name="Dirichlet-Mesh">
```
to follow the convention. Similar below.

BenjaminRodenberg replied:
Ok. I will commit this change and clean up everything accordingly.

uekerman commented on partitioned-heat-conduction-complex/precice-config.xml:
```diff
@@ -0,0 +1,61 @@
+    <coupling-scheme:serial-implicit>
+      <participants first="HeatDirichlet" second="HeatNeumann"/>
+      <max-time value="1"/>
+      <time-window-size value=".1" valid-digits="8"/>
```
```suggestion
      <time-window-size value=".1" />
```

uekerman commented on partitioned-heat-conduction-complex/precice-config.xml:
```diff
@@ -0,0 +1,61 @@
+      <use-mesh name="NeumannNodes" from="HeatNeumann"/>
+      <write-data name="Flux" mesh="DirichletNodes"/>
+      <read-data name="Temperature" mesh="DirichletNodes"/>
+      <mapping:nearest-projection direction="read" from="NeumannNodes" to="DirichletNodes" constraint="consistent" timing="initial"/>
```
```suggestion
      <mapping:nearest-projection direction="read" from="NeumannNodes" to="DirichletNodes" constraint="consistent" />
```

uekerman commented on partitioned-heat-conduction-complex/precice-config.xml:
```diff
@@ -0,0 +1,61 @@
+      <use-mesh name="DirichletNodes" from="HeatDirichlet"/>
+      <write-data name="Temperature" mesh="NeumannNodes"/>
+      <read-data name="Flux" mesh="NeumannNodes"/>
+      <mapping:nearest-projection direction="read" from="DirichletNodes" to="NeumannNodes" constraint="consistent" timing="initial"/>
```
```suggestion
      <mapping:nearest-projection direction="read" from="DirichletNodes" to="NeumannNodes" constraint="consistent" />
```

uekerman commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
 summary:
 ---
 
-{% include important.html content="We have not yet ported the documentation of the preCICE tutorials from the preCICE wiki to here. Please go to the [preCICE wiki](https://github.com/precice/precice/wiki#2-getting-started---tutorials)" %}
```
We still need this. Otherwise the website will already show wrong information. Also please add this to the complex case.

BenjaminRodenberg replied:
Resolved in e1d1411

uekerman commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
 permalink: tutorials-partitioned-heat-conduction.html
-keywords:
+keywords: FEniCS, Nutils, Heat conduction
 summary:
```
TODO

BenjaminRodenberg replied:
What exactly do you mean?

uekerman replied:
To write a summary

BenjaminRodenberg replied:
Ok. I was totally focused on the diff. This makes sense :sweat_smile:

BenjaminRodenberg replied:
Done. bf765be1be7216f28d8ccabe7d04036dae8627eb

uekerman commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
-{% include important.html content="We have not yet ported the documentation of the preCICE tutorials from the preCICE wiki to here. Please go to the [preCICE wiki](https://github.com/precice/precice/wiki#2-getting-started---tutorials)" %}
+## Setup
+
+In this tutorial we solve a partitioned heat equation. For information on the non-partitioned case, please refer to [1, p.37ff]. The corresponding FEniCS code can be found on [github:fenics-tutorial](https://github.com/hplgit/fenics-tutorial/blob/master/pub/python/vol1/ft03_heat.py).
```
Needs probably a bit more information. We have not yet mentioned FEniCS before.

BenjaminRodenberg replied:
I provided more details on the case in 75e68af. I don't think that we have to explain what FEniCS is here. Or what do you mean?

uekerman replied:
Right now, "FEniCS" falls a bit from the sky. The setup should actually be independent of the solver.

BenjaminRodenberg replied:
I did some reordering in 9614c02. I think this makes it a bit better.

uekerman commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
+
+* `fenics`, requires you to install [FEniCS](https://fenicsproject.org/download/) and the [FEniCS-adapter](https://github.com/precice/fenics-adapter). 
+
+* :construction: This case is still under construction. See https://github.com/precice/tutorials/issues/152. :construction: `nutils`, requires you to install [Nutils](http://www.nutils.org/en/latest/).
```
The "construction" is not necessary. Everything is still under construction :smile: 
The case already runs and gives qualitatively good results, right?

BenjaminRodenberg replied:
The nutils case does not give good result yet. See #152. Therefore I would like to merge this one and mark the nutils case as "under construction".

uekerman replied:
What does "not good" exactly mean? :grin:

BenjaminRodenberg replied:
See below: I removed the `closes #137` to make this PR less confusing. Will work on the rest in #152

uekerman commented on partitioned-heat-conduction/README.md:
```diff
@@ -1,8 +1,58 @@
+
+## Visualization
+
+Output is written into the folder `fenics/out`. You can visualize the content with paraview by opening the `*.pvd` files. The files `HeatDirichlet` and `HeatNeumann` correspond to the numerical solution of the Dirichlet, respectively Neumann, problem, while the files with the prefix `ref` correspond to the analytical reference solution.
+
```
Nutils missing.

BenjaminRodenberg replied:
I would prefer solving this in #152, as well.

uekerman replied:
OK, but then let's maybe remove that this PR solves #137 
Otherwise, I get too confused :smile:

BenjaminRodenberg replied:
I removed the `closes #137` to make this PR less confusing. Will work on the rest in #152

uekerman reviewed (changes requested):
Overall structure looks good. Minor conventions things and README stuff below.

uekerman commented:
> Currently we don't provide `fenics/run.sh`, since we need command line input to decide whether we run the Dirichlet or Neumann solver. I see two options here:
> 
>     1. provide a `fenics/run.sh` that accepts command line input
> 
>     2. provide a `dirichlet-fenics/run.sh` and `neumann-fenics/run.sh`. Put all actual code only into `dirichlet-fenics`. `neumann-fenics` only exists for following the general structure and providing a `run.sh` script.
> 
> 
> Any opinion?

I guess we are also fine without a run script here for the moment. If we add some later then option 1.

uekerman reviewed (approved):
Looks good. Did not run it again, but did not see any obvious problems.
